upup/pkg/fi/cloudup/urls.go

```diff
@@ -17,74 +17,162 @@ limitations under the License.
 package cloudup
 
 import (
+	"net/url"
 	"os"
 	"strings"
 
+	"path"
+
 	"github.com/golang/glog"
 	"k8s.io/kops"
+	"k8s.io/kops/pkg/assets"
+	"k8s.io/kops/util/pkg/hashing"
 )
 
-// baseUrl caches the BaseUrl value
-var baseUrl string
+var kopsUrl = "https://kubeupv2.s3.amazonaws.com/kops/" + kops.Version + "/"
+var baseUrl *url.URL
 
 // BaseUrl returns the base url for the distribution of kops - in particular for nodeup & docker images
-func BaseUrl() string {
-	if baseUrl != "" {
+func BaseUrl() (*url.URL, error) {
+	if baseUrl != nil {
 		// Avoid repeated logging
-		return baseUrl
+		return baseUrl, nil
 	}
 
-	baseUrl = os.Getenv("KOPS_BASE_URL")
-	if baseUrl == "" {
-		baseUrl = "https://kubeupv2.s3.amazonaws.com/kops/" + kops.Version + "/"
-		glog.V(4).Infof("Using default base url: %q", baseUrl)
+	baseUrlString := os.Getenv("KOPS_BASE_URL")
+	if baseUrlString == "" {
+		baseUrlString = kopsUrl
+		glog.V(4).Infof("Using default base url: %q", kopsUrl)
 	} else {
-		glog.Warningf("Using base url from KOPS_BASE_URL env var: %q", baseUrl)
+		// TODO I do not think we need to do this with
+		if !strings.HasSuffix(baseUrlString, "/") {
+			baseUrlString += "/"
+		}
+		glog.Infof("Using base url from KOPS_BASE_URL env var: %q", baseUrlString)
 	}
 
-	if !strings.HasSuffix(baseUrl, "/") {
-		baseUrl += "/"
+	baseUrl, err := url.Parse(baseUrlString)
+	if err != nil {
+		return nil, err
 	}
 
-	return baseUrl
+	return baseUrl, nil
 }
 
-// nodeUpLocation caches the NodeUpLocation value
-var nodeUpLocation string
+// SetKopsAssetsLocations sets the kops assets locations
+func SetKopsAssetsLocations(assetsBuilder *assets.AssetBuilder) error {
+	for _, s := range []string{
+		"linux/amd64/kops", "darwin/amd64/kops",
+	} {
+		_, _, err := FileUrl(s, assetsBuilder)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+var nodeUpLocation *url.URL
+var nodeUpHash *hashing.Hash
 
 // NodeUpLocation returns the URL where nodeup should be downloaded
-func NodeUpLocation() string {
-	if nodeUpLocation != "" {
+func NodeUpLocation(assetsBuilder *assets.AssetBuilder) (*url.URL, *hashing.Hash, error) {
```

Comment:
Both the nodeup and protokube functions are very alike. They use the default, the ENV variable, or the asset location of that file.  That file and the hash are cached in variables.  Previously we were using strings, and this refactors to URLs and introduces the asset builder.